File: 338-counting-bits/338-counting-bits.scala
object Solution {
    def countBits(n: Int): Array[Int] = {
        val res: Array[Int] = Array.fill(n + 1) {0}
        var pow2 = 1
        var j = 0
        
        for(i <- 1 to n) {
            if (i == pow2) {
                pow2 *= 2
                j = 0
            }
            
            res(i) = res(j) + 1
            j += 1
        }
        
        res
    }
}

File: 338-counting-bits/338-counting-bits.java
class Solution {
    public int[] countBits(int n) {
        int[] res = new int[n + 1];
        
        for(int i = 1, cur = 1, p = 0; i <= n; i++, p++) {
            if (i == cur) {
                cur = cur * 2;
                res[i] = 1;
                p = 0;
                continue;
            }
            res[i] = res[p] + 1;
        }
        
        return res;
    }
}